core/src/main/java/org/infinispan/transaction/impl/TransactionTable.java

```diff
@@ -369,20 +369,31 @@ private RemoteTransaction getOrCreateRemoteTransaction(GlobalTransaction globalT
          throw log.cacheIsStopping(cacheName);
       }
 
-      remoteTransaction = modifications == null ? txFactory.newRemoteTransaction(globalTx, topologyId)
-            : txFactory.newRemoteTransaction(modifications, globalTx, topologyId);
-      RemoteTransaction existing = remoteTransactions.putIfAbsent(globalTx, remoteTransaction);
-      if (existing != null) {
-         if (trace) log.tracef("Remote transaction already registered: %s", existing);
-         return existing;
-      } else {
-         if (trace) log.tracef("Created and registered remote transaction %s", remoteTransaction);
-         if (remoteTransaction.getTopologyId() < minTxTopologyId) {
-            if (trace) log.tracef("Changing minimum topology ID from %d to %d", minTxTopologyId, remoteTransaction.getTopologyId());
-            minTxTopologyId = remoteTransaction.getTopologyId();
+      return remoteTransactions.compute(globalTx, (gtx, existing) -> {
+         if (existing != null) {
+            if (trace)
+               log.tracef("Remote transaction already registered: %s", existing);
+            return existing;
+         } else {
+            if (isTransactionCompleted(globalTx)) {
+               throw new IllegalStateException("Cannot create remote transaction " + globalTx + ", already completed");
+            }
+            if (!rpcManager.getTransport().getMembers().contains(gtx.getAddress())) {
```

Comment:
Good catch, I was incorrectly assuming that the iteration in `cleanupLeaverTransactions()` would block if an entry was locked by `compute()`. I'm moving the originator check after `compute()`.